### PR TITLE
PYTHON-2963 Add tox config in preparation for migration from setup.py

### PR DESCRIPTION
**Pull request by NoahStapp**
This change is still using `python setup.py test` inside tox to run the unit tests: migration to pytest will be done in https://jira.mongodb.org/browse/PYTHON-3727, and migration to pyproject.toml will be done in https://jira.mongodb.org/browse/PYTHON-2965.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,45 @@
+[tox]
+requires =
+    tox>=4
+envlist =
```
Comment (NoahStapp) [inline tox.ini:4]: The lack of specified Python versions here will cause tox to run using the default system interpreter.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    pre-commit run --all-files
+
+[testenv:lint-manual]
+description = run all pre-commit stages, including those that require manual fixes
+deps =
+    pre-commit
+commands =
+    pre-commit run --all-files
+    pre-commit run --all-files --hook-stage manual flake8
+    pre-commit run --all-files --hook-stage manual doc8
```
Comment (ShaneHarvey) [inline tox.ini:34]: `pre-commit run --all-files --hook-stage manual` will run all the pre-commit checks in one command.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
If the plan is to migrate to tox in future PRs then could you update the title to something like "Add tox config"?

File: tox.ini (hunk 2 above)
Comment (NoahStapp) [reply tox.ini]: Good catch, thanks!

**Review by ShaneHarvey — APPROVED** [review]
LGTM!

**Review by blink1073 — CHANGES_REQUESTED** [review]
We should update the `github` workflow as part of this PR to use `tox`.

**Comment by blink1073** [comment]
Note that moving our evergreen tests to `tox` will involve adding new environments for things like `encryption`, etc, and move things out of the EG files like https://github.com/mongodb/mongo-python-driver/blob/master/.evergreen/run-tests.sh.   @NoahStapp, can you please add a note to the Epic that we'll need tickets for each of those files to migrate to tox?

**Comment by NoahStapp** [comment]
> Note that moving our evergreen tests to `tox` will involve adding new environments for things like `encryption`, etc, and move things out of the EG files like [`master`/.evergreen/run-tests.sh](https://github.com/mongodb/mongo-python-driver/blob/master/.evergreen/run-tests.sh?rgh-link-date=2023-06-15T18%3A54%3A18Z). @NoahStapp, can you please add a note to the Epic that we'll need tickets for each of those files to migrate to tox?

Sure, will do.

**Comment by NoahStapp** [comment]
> We should update the `github` workflow as part of this PR to use `tox`.

For all of the workflow jobs?

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+description = run pyright to typecheck
+deps =
+    mypy
+    pyright==1.1.290
+allowlist_externals=echo
+commands =
+    echo '{"strict": ["tests/test_typing_strict.py"]}' >> pyrightconfig.json
+    pyright test/test_typing_strict.py
+
+[testenv:typecheck]
```
Comment (blink1073) [inline tox.ini:69]: I don't think we need this environment, we can call the individual ones.

Comment (NoahStapp) [reply tox.ini:69]: This is intended to replace needing an alias to run all of the typechecks in a single command.

Comment (blink1073) [reply tox.ini:69]: Hmm, it is possible then to compose this to use the deps and commands of the other envs?

Comment (NoahStapp) [reply tox.ini:68]: Figured it out, fixed!

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+commands =
+    pyright test/test_typing.py test/test_typing_strict.py
+
+[testenv:typecheck-pyright-strict]
+description = run pyright with strict mode to typecheck
+deps =
+    {[testenv:typecheck-pyright]deps}
+allowlist_externals=echo
+commands =
+    echo '{"strict": ["tests/test_typing_strict.py"]}' >> pyrightconfig.json
```
Comment (ShaneHarvey) [inline tox.ini:65]: `>>` should be `>` IIUC. Otherwise we'd repeatedly append to the same file.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

**Review by blink1073 — APPROVED** [review]
Thanks!

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**